### PR TITLE
Update commonmark to v2.0

### PR DESCRIPTION
**Pull request by gagahpangeran**
- [x] depends on #8091

In this PR:

- updating commonmark from v1.6 to v2.0.
- Change almost all namespace and class because so many breaking changes.
- add more markdown test.
- separate config in `OsuMarkdown`. Because in v2.0 config schema needs to be declared, I decided to separate the config.
- use `DefaultAttributesExtension` to add class name and other simple attributes. This reduce complexity in `DocumentProcessor` by removing method `addClass` and `addListStartAsVariable`, and also support type hint when accessing the node via its callable. (can be revert if unnecessary in this PR)

After back and forth reading breaking changes documentation and also dive into commonmark code base, finally it's here!

There is so much breaking changes in this version, from API changes to namespace changes.

Reference to read:
- https://commonmark.thephpleague.com/2.0/upgrading/developers/
- https://commonmark.thephpleague.com/2.0/changelog/

I hope I didn't miss something.

### CODE REVIEW

**Comment by nanaya** [comment]
How is this supposed to be merged if this breaks wiki page...?

**Comment by nanaya** [comment]
Also it's probably better to update existing parser to support both `{{{` and `:::` syntax so existing wiki pages can be updated ahead of time.

**Comment by gagahpangeran** [comment]
> How is this supposed to be merged if this breaks wiki page...?

Which part is break?

I already test on local and seems working.

> Also please separate framework upgrade in a different pr even if it doesn't involve other code changes.

Okay, will do.

> Also it's probably better to update existing parser to support both `{{{` and `:::` syntax so existing wiki pages can be updated ahead of time.

I decided to rewrite the `StyleBlock` with different syntax because currently there are only two pages in wiki that use `StyleBlock`, so it can be easily changed after deploy.

cc : @TicClick to confirm.

**Comment by nanaya** [comment]
> > How is this supposed to be merged if this breaks wiki page...?
> 
> Which part is break?
> 

The part which StyleBlock is gone.

> > Also it's probably better to update existing parser to support both `{{{` and `:::` syntax so existing wiki pages can be updated ahead of time.
> 
> I decided to rewrite the `StyleBlock` with different syntax because currently there are only two pages in wiki that use `StyleBlock`, so it can be easily changed after deploy. But I could try to support both if it's necessary for now.
> 

This PR may take a while to be merged so update the existing one to support both so wiki pages can be updated first.

And don't bother separating the PR as they should be part of the update.

**Comment by gagahpangeran** [comment]
Because current use case in wiki seems there is no need to support nested StyleBlock, I guess I can ignore the nested case with `{{{` syntax for now.? Because it's a bit hard with `{{{` syntax to handle nested block for now.

**Comment by nanaya** [comment]
Just remove the nested syntax support altogether for now (or ever, until it's actually needed).

**Comment by TicClick** [comment]
thanks for the cc. it's hard for me to imagine a scenario where we would really need nested custom blocks, so you're absolutely free to skip this part.

for the period of transition, it doesn't matter much. the only two pages that use infoboxes right now are Modding/Nuke (low traffic) and Silence. I can coin a PR with the syntax replacement really quick, and someone would merge it once the update is deployed (@nanaya I think all of the web develops with write access also have control over the wiki repository, so this part is covered)

**Comment by nanaya** [comment]
I have feeling I didn't phrase it clearly but this should be the steps:

1. create a different PR updating 1.x markdown extension to support `:::` (with test etc)
2. update wiki once `1.` is merged
3. continue with this PR

**Comment by gagahpangeran** [comment]
> I have feeling I didn't phrase it clearly but this should be the steps:
> 
>     1. create a different PR updating 1.x markdown extension to support `:::` (with test etc)
> 
>     2. update wiki once `1.` is merged & deployed
> 
>     3. continue with this PR (only supporting `:::`)

Ah, okay. sorry for the misunderstanding.

File: app/Libraries/Markdown/OsuMarkdown.php [hunk 1]
```diff
@@ -203,23 +267,94 @@ private function getIndexableConverter(): MarkdownConverter
...
+            $config = array_merge(
+                $config,
+                [
+                    'osu_extension' => $this->osuExtensionConfig,
+                    'default_attributes' => $this->createDefaultAttributesConfig(),
+                ]
+            );
+        }
+
+        $environment = new Environment($config);
```
Comment (nanaya) [inline app/Libraries/Markdown/OsuMarkdown.php:284]: the additional config and the `DefaultAttributesExtension` below belong in `getHtmlConverterAndExtension`

```
$environment = new Environment(array_merge($this->commonmarkConfig, $extraConfig));
```

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Libraries/Markdown/Osu/Renderers/ListItemRenderer.php [hunk 2]
```diff
@@ -5,27 +5,29 @@
...
-    public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, bool $inTightList = false)
+    public function render(Node $node, ChildNodeRendererInterface $childRenderer)
     {
-        $li = $this->baseRenderer->render($block, $htmlRenderer, $inTightList);
+        $li = $this->baseRenderer->render($node, $childRenderer);
 
-        $contents = $li->getContents();
-
-        $li->setContents("<div>{$contents}</div>");
+        if ($li instanceof HtmlElement) {
```
[the remark this answers is not in the document]
Comment (gagahpangeran) [reply app/Libraries/Markdown/Osu/Renderers/ListItemRenderer.php:27]: Sure, what error should be thrown?

[the remark this answers is not in the document]
Comment (nanaya) [reply app/Libraries/Markdown/Osu/Renderers/ListItemRenderer.php:27]: `InvalidArgumentException` should be fine.

**Comment by nanaya** [comment]
composer lockfile need update

```
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
```

File: app/Libraries/Markdown/Osu/DocumentProcessor.php [hunk 3]
```diff
@@ -6,35 +6,37 @@
...
+    private array $tocSlugs;
 
-    private $relativeUrlRoot;
-    private $wikiLocale;
-    private $wikiPathToRoot;
-    private $wikiAbsoluteRootPath;
+    private ?string $relativeUrlRoot;
+    private ?string $wikiLocale;
+    private string $wikiPathToRoot;
+    private string $wikiAbsoluteRootPath;
```
Comment (nanaya) [inline app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: this causes error on news page `/home/news/2021-06-02-new-featured-artist-fleshgod-apocalypse` (or comment containing link like `[link label](/wiki/page)`)

though looking again the function probably shouldn't even be run on non-wiki pages in the first place

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Libraries/Markdown/Osu/DocumentProcessor.php (hunk 3 above)
Comment (gagahpangeran) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: Ah, I missed this.

So `$wikiPathToRoot` and `$wikiAbsoluteRootPath` are nullable?

> though looking again the function probably shouldn't even be run on non-wiki pages in the first place

Which function is this?

Comment (nanaya) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: `fixWikiUrl`

**Comment by nanaya** [comment]
Not sure if want to update a bunch of unrelated libraries here

**Comment by gagahpangeran** [comment]
> Not sure if want to update a bunch of unrelated libraries here

Hmm, I only run `composer update`, not sure what happened :thinking:

**Comment by nanaya** [comment]
`composer update` updates everything. Better way is to reinstall the added package using the `require` command

File: app/Libraries/Markdown/Osu/DocumentProcessor.php (hunk 3 above)
Comment (gagahpangeran) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: Is there any way to check whether it is wiki page or not?

Comment (nanaya) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: Checking the relevant config (the same as the function which sets the variable) should be enough although I noticed it's still kinda needed for news posts

Comment (gagahpangeran) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: Yeah, wiki link also need fixed in news posts.

Comment (nanaya) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php:35]: it's fine for the links. It's just the images take one extra redirect when loading the page.

Comment (gagahpangeran) [reply app/Libraries/Markdown/Osu/DocumentProcessor.php]: Done.

**Review by nanaya — APPROVED** [review]

**Pull request merged**